packages/versioning/src/lib.rs

```diff
@@ -1,115 +1,187 @@
+        if saved_version.software < version.software {
+            Ok(version)
+        } else {
+            Err(StdError::generic_err(
+                "Couldn't upgrade contract because version isn't monotonically increasing!",
+            ))
+        }
+    })?;
 
-    const QUERY_VERSION: WithVersion<VariantsQuery> =
-        WithVersion::new_query(VariantsQuery::Version {});
+    Ok(())
+}
 
-    fn assert_query_serde(value: WithVersion<VariantsQuery>) {
-        assert_eq!(
-            serde_json::from_str::<WithVersion<VariantsQuery>>(
-                &serde_json::to_string(&value).unwrap()
-            )
-            .unwrap(),
-            value
-        );
+pub fn update_software_and_storage<
+    'r,
+    const FROM_STORAGE_VERSION: VersionSegment,
+    MigrateStorageFunctor,
+    MigrateStorageError,
+>(
+    storage: &'r mut dyn Storage,
+    version: Version,
+    migrate_storage: MigrateStorageFunctor,
+) -> Result<(), MigrateStorageError>
+where
+    MigrateStorageFunctor: FnOnce(&'r mut dyn Storage) -> Result<(), MigrateStorageError>,
+    MigrateStorageError: From<StdError> + Error,
+{
+    if version.storage == FROM_STORAGE_VERSION {
+        return Err(StdError::generic_err("Software and storage update handler called, but expected and new storage versions are the same!").into());
     }
 
-    #[test]
-    fn test_query_serde() {
-        assert_query_serde(VERSION);
-        assert_query_serde(QUERY_ABC);
-        assert_query_serde(QUERY_DEF);
-        assert_query_serde(QUERY_VERSION);
-
-        assert_eq!(
-            serde_json::from_str::<WithVersion<VariantsQuery>>(r#"{"version":null}"#).unwrap(),
-            VERSION
-        );
-
-        assert_eq!(
-            serde_json::from_str::<WithVersion<VariantsQuery>>(r#"{"abc":{}}"#).unwrap(),
-            QUERY_ABC
-        );
-
-        assert_eq!(
-            serde_json::from_str::<WithVersion<VariantsQuery>>(r#"{"def":{}}"#).unwrap(),
-            QUERY_DEF
-        );
-
-        assert_eq!(
-            serde_json::from_str::<WithVersion<VariantsQuery>>(r#"{"version":{}}"#).unwrap(),
-            QUERY_VERSION
-        );
+    if version.storage != FROM_STORAGE_VERSION.wrapping_add(1) {
+        return Err(StdError::generic_err("Expected and new storage versions are not directly adjacent! This could indicate an error!").into());
     }
+
+    VERSION_STORAGE_KEY.update(storage, |saved_version| {
+        if saved_version.storage != version.storage {
+            return Err(StdError::generic_err(
+                "Couldn't upgrade contract because saved storage version didn't match expected one!",
+            ));
+        }
+
+        if saved_version.software < version.software {
+            Ok(version)
+        } else {
+            Err(StdError::generic_err(
+                "Couldn't upgrade contract because software version isn't monotonically increasing!",
+            ))
+        }
```

Comment:
I believe we may re-use this code fragment with  `update_software` .... and limit the bugs :-)

Reply:
Addressed.